contracts/lib/HashLib.sol

```diff
@@ -150,20 +175,36 @@ library HashLib {
         pure
         returns (bytes32 _hash)
     {
-        // chainId is not needed as it is in the ChainSession
-        _hash = keccak256(
+        {
+            // chainId is not needed as it is in the ChainSession
+            _hash = keccak256(
+                abi.encode(
+                    SESSION_TYPEHASH,
+                    account,
+                    smartSession,
+                    uint8(mode), // Include mode as uint8
+                    address(session.sessionValidator),
+                    session.salt,
+                    keccak256(session.sessionValidatorInitData),
+                    hashPermissions(session),
+                    nonce
+                )
+            );
+        }
+    }
+
+    function hashPermissions(Session memory session) internal pure returns (bytes32) {
```

Comment:
@filmakarov @kopy-kat I had to move these data points into another EIP712 envelope and function to avoid stack too deep